source/common/http/codec_client.h

```diff
@@ -223,7 +224,9 @@ class CodecClient : protected Logger::Loggable<Logger::Id::client>,
                          public ResponseDecoderWrapper,
                          public RequestEncoderWrapper {
     ActiveRequest(CodecClient& parent, ResponseDecoder& inner)
-        : ResponseDecoderWrapper(inner), RequestEncoderWrapper(nullptr), parent_(parent) {
+        : ResponseDecoderWrapper(inner), RequestEncoderWrapper(nullptr), parent_(parent),
+          header_validator_(
+              parent.host_->cluster().makeHeaderValidator(parent.codec_->protocol())) {
```

Comment:
each client will get their own validator?  We probably do this on the downstream side but is this expensive? Should we instead have one shared across the cluster, and call validate() with the codec, or is it largely stateless so cheap?

Reply:
They will need to have each own validator to support upgrade/connect normalization. It is not expensive. Right now UHV has no state at all. Once I move upgrade/connect normalization there, it will have state for that. But it is no more compared to the non-UHV case.